src/mlir/design.md

```diff
@@ -3,69 +3,225 @@
+We could add annotations to the lowered code (potentially multiple instructions) with metadata, but that would require us to search and bundle instructions every time when looking for high-level patterns.
 
 For those reasons, Verona will need a light dialect to represent type constraints, region information and module boundaries, which don't naturally convert to LLVM IR.
+This will allow us to do high-level optimisatins cheaply, before we lower that code to LLVM.
 
 ## Generation Simplicity
 
-The new AST is much simpler than the previous.
-The new parser uses a lean typed AST and converts all language into those nodes, which becomes much simpler to lower to an IR.
-The parser also runs the type checks and inference, so the AST we get at this stage is already complete (explicit concrete monomorphic types, reified code only), and checked for syntactic and semantic errors.
+The new AST is much simpler than the previous. 
+It has only primitive control flow (`assign`, `select`, `match`, `when`, etc), from which higher level control flow is constructed (conditionals, loops, calls, member access, etc).
+It also has support strucures with all reachable types and functions, which will be the main driver of the MLIR lowering.
 
-If an error is detected when generating MLIR, however, we still have the original source locations and can return the errors to the parser's own diagnostics infrastructure.
+The parser runs the type checks and inference, so the information we get at this stage is already complete (explicit concrete monomorphic types, reified code only), and checked for syntactic and semantic errors.
 
-The MLIR dialect that we'll develop at this stage will be solely for representing the lowering concepts in order to simplify high-level optimisations before lowering to LLVM IR.
-In theory, the AST could have been directly lowered to LLVM IR, but the LLVM pipeline wouldn't be able, for example, to optimise based on type and region information, memory management mechanisms (ex. ref count), object layout and representation, etc.
+Verona also doesn't restrict declaration order. Functions and classes can be declared at any time in the module to be used anywhere.
+To avoid order issues in the IR, we'll do multiple passes on the structures:
+1. Declaration pass: Defines reachable type structures and functions, including the entire standard library files.
+2. Lowering pass: Lower the reachable AST nodes using the structures above and defining the functions for the prototypes above.
+
+In theory, the AST could have been directly lowered to LLVM IR, but the LLVM IR is too low level for some analysis we want to do (ex. region aliasing) and it would be counter-productive to do so at that stage.
+
+## Classes & Modules
+
+Classes and modules are the same concept in Verona.
+Each source sub-directory containing Verona files is a module, and all its Verona files share the same namespace, named after the directory name.
```

Comment:
This is only half right.  There are two names for a module.  The name it shows up under in the source file where it is used and the name that tells the compiler where to find it.  I presume the AST handles the combination of these for you?

I also presume anything about visibility is handled before we get to this level?

Reply:
Precisely.